src/ChainRouter.php

```diff
@@ -227,8 +231,19 @@ public function generate($name, $parameters = [], $absolute = UrlGeneratorInterf
                 continue;
             }
 
+            // if $router does not announce it is capable of handling
+            // non-string routes and the ROUTE_OBJECT is set in the parameters array, continue
+            if (array_key_exists(RouteObjectInterface::ROUTE_OBJECT, $parameters) && is_object($parameters[RouteObjectInterface::ROUTE_OBJECT]) && !$router instanceof VersatileGeneratorInterface) {
+                continue;
+            }
+
+            $routeName = $name;
+            if (RouteObjectInterface::OBJECT_BASED_ROUTE_NAME === $name && array_key_exists(RouteObjectInterface::ROUTE_OBJECT, $parameters) && is_object($parameters[RouteObjectInterface::ROUTE_OBJECT])) {
+                $routeName = $parameters[RouteObjectInterface::ROUTE_OBJECT];
```

Comment:
i would not do this. instead i would only call the versatile generator `supports` if $name is an object, and deprecate the interface. we had this because normal routers choke on $name being an object. with the new way of doing this, $name won't be an object anymore and the router can just do a 404.